vochain/process_id.go

```diff
@@ -0,0 +1,120 @@
+// - census proof type: kind of census used for the process (from protobuf models)
+//
+// - nonce: an incremental uint32 number
+//
+// processID: 395b41cc9abcd8da6bf26964af9d7eed9e03e53415d37aa96045000600000001 represents
+// chID:vocdoni-azeno addr:0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045 pt:arbo nc:1
+type ProcessID struct {
+	chainID          string
+	organizationAddr *common.Address
+	proofType        uint16
+	nonce            uint32
+}
+
+// Marshal encodes to bytes
+func (p *ProcessID) Marshal() []byte {
+	//processID = chainID[6bytes] + organizationAddress[20bytes] + proofType[2bytes] + nonce[4bytes]
+	chainID := ethereum.HashRaw([]byte(p.chainID))
+
+	nonce := make([]byte, 4)
+	binary.BigEndian.PutUint32(nonce, uint32(p.nonce))
+
+	proofType := make([]byte, 2)
+	binary.BigEndian.PutUint16(proofType, p.proofType)
+
+	id := make([]byte, 32)
+	if p.organizationAddr == nil {
+		panic("organization cannot be nil")
+	}
+	id = append(chainID[:6], p.organizationAddr.Bytes()[:20]...)
+	id = append(id, proofType[:2]...)
+	return append(id, nonce[:4]...)
+}
+
+// Unmarshal decodes a 32 byte payload into the process ID fields
+func (p *ProcessID) Unmarshal(pid []byte) error {
+	if len(pid) != 32 {
+		return fmt.Errorf("process ID lenght not correct")
+	}
+	p.chainID = fmt.Sprintf("%x", pid[:4])
+	baddr := make([]byte, 20)
+	copy(baddr[:], pid[6:26])
+	addr := common.BytesToAddress(baddr)
+	p.organizationAddr = &addr
+	p.proofType = binary.BigEndian.Uint16(pid[26:28])
+	p.nonce = binary.BigEndian.Uint32(pid[28:32])
+	return nil
+}
+
+// String returns a human readable representation of process ID
+func (p *ProcessID) String() string {
+	return fmt.Sprintf("[chID:%s/addr:%s/pt:%s/nc:%d]",
+		p.chainID, p.organizationAddr.Hex(), p.Proof(), p.nonce)
+
+}
+
+// SetChainID sets the blockchain identifier
+func (p *ProcessID) SetChainID(chID string) {
+	p.chainID = chID
+}
+
+// ChainID returns blockchain identifier
+func (p *ProcessID) ChainID() string {
+	return p.chainID
+}
+
+// SetAddr sets the organizer address (20 bytes)
+func (p *ProcessID) SetAddr(addr common.Address) {
+	p.organizationAddr = &addr
+}
+
+// Addr returns the organizer address
+func (p *ProcessID) Addr() *common.Address {
+	return p.organizationAddr
+}
+
+// SetNonce sets the nonce number that identifies the process
+func (p *ProcessID) SetNonce(nonce uint32) {
+	p.nonce = nonce
+}
+
+// Nonce returns the process ID identifier number
+func (p *ProcessID) Nonce() uint32 {
+	return p.nonce
+}
+
+// SetProof sets the census proof type
+func (p *ProcessID) SetProof(proof *models.Proof) error {
+	if proof == nil {
+		return fmt.Errorf("proof is nil")
+	}
+	// get the Protobuf oneOf descriptor index number and use it as proofType
+	protoDesc := proof.ProtoReflect().Descriptor().Oneofs().Get(0)
+	p.proofType = uint16(proof.ProtoReflect().WhichOneof(protoDesc).Number())
+	return nil
+}
+
+// Proof returns a string representing the census proof type
+func (p *ProcessID) Proof() string {
```

Comment:
nit: SetProof takes a pointer, but Proof takes a string. would be nice for the two to be consistent.